ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+                        std::vector<int64_t> unsqueeze_axis1;
+                        std::vector<int64_t> unsqueeze_axis2;
+                        for (const auto& sep_label2 : separate_labels2)
+                        {
+                            NGRAPH_CHECK(label_to_dim_map2.find(sep_label2) !=
+                                         label_to_dim_map2.end());
+                            auto label_dims = label_to_dim_map2[sep_label2];
+                            for (size_t dim_ind = 0; dim_ind < label_dims.size(); ++dim_ind)
+                            {
+                                unsqueeze_axis1.push_back(unsqueeze_dim +
+                                                          static_cast<int64_t>(dim_ind));
+                            }
+                        }
+                        for (const auto& sep_label1 : separate_labels1)
+                        {
+                            NGRAPH_CHECK(label_to_dim_map1.find(sep_label1) !=
+                                         label_to_dim_map1.end());
+                            auto label_dims = label_to_dim_map1[sep_label1];
+                            for (auto label_dim : label_dims)
+                            {
+                                unsqueeze_axis2.push_back(label_dim);
+                            }
+                        }
+
+                        // unsqueeze input operands for elementwise-multiplication with broadcasting
+                        HostTensorPtr unsqueeze_output1 =
+                            unsqueeze_input<T>(input1, unsqueeze_axis1);
+                        HostTensorPtr unsqueeze_output2 =
+                            unsqueeze_input<T>(input2, unsqueeze_axis2);
+
+                        // multiply both operands with broadcasting
+                        PartialShape output_shape = unsqueeze_output1->get_partial_shape();
+                        PartialShape::broadcast_merge_into(output_shape,
+                                                           unsqueeze_output2->get_partial_shape(),
+                                                           ngraph::op::AutoBroadcastSpec::NUMPY);
+                        HostTensorPtr mul_output = std::shared_ptr<HostTensor>(new HostTensor(
+                            unsqueeze_output1->get_element_type(), output_shape.get_shape()));
+                        ngraph::runtime::reference::multiply<T>(
+                            unsqueeze_output1->get_data_ptr<T>(),
+                            unsqueeze_output2->get_data_ptr<T>(),
+                            mul_output->get_data_ptr<T>(),
+                            unsqueeze_output1->get_shape(),
+                            unsqueeze_output2->get_shape(),
+                            ngraph::op::AutoBroadcastSpec::NUMPY);
+
+                        // update input operand and input subscript for Einsum operation
+                        update_operands(inputs,
```

Comment:
If you update inputs, looks like in some method which were called before, we can rewrite data in input tensors instead of new tensor creation.

Reply:
this method does not create new tensor. it just updates pointers in a vector.

Reply:
try to re-use already allocated memory for inter. results

Reply:
Done.